# Práctica: Análisis de Datos y Clustering (Auto MPG)

 * **Autor:** Jolehisy Acevedo Medina
 * **Fecha:** 25/03/2026
 * **Asigantura:** Informática Industrial Avanzada
 * **Máster en Ingeniería Industrial**
 * **Universidad de La Laguna**

## Descripción del Dataset
Para esta tarea vamos a utilizar el dataset **"California Housing"**, que contiene datos del censo sobre los distritos de California. Es un dataset que combina variables continuas y categóricas.

Entre las variables destacan:
* **housing_median_age**: Edad media de las casas del distrito.
* **total_rooms** / **total_bedrooms**: Total de habitaciones y dormitorios.
* **median_income**: Ingreso medio de los residentes.
* **ocean_proximity**: Ubicación respecto al mar (Variable categórica de texto).

## Objetivos del Proyecto
El objetivo de este cuaderno es aplicar técnicas de Inteligencia Artificial y Análisis de Datos industrial siguiendo estos pasos:
1. **Análisis Exploratorio (EDA)**: Cargar los datos y revisar su estructura inicial.
2. **Limpieza y Preparación**: Localizar valores nulos (datos faltantes), tratarlos adecuadamente y convertir las variables de texto a formato numérico para que la IA pueda procesarlas.
3. **Clustering (Aprendizaje No Supervisado)**: Aplicar algoritmos como *K-Means* y *Gaussian Mixture Models (GMM)* para que la IA agrupe automáticamente los vehículos según sus similitudes técnicas.
4. **Evaluación y Comparación**: Reducir las dimensiones para visualizar los clústeres generados y comparar el rendimiento de ambos algoritmos.

### 1. Importación de Librerías y Carga de Datos
El primer paso en cualquier proyecto de análisis de datos es importar las herramientas necesarias. Utilizaremos **Pandas**, que es la librería estándar en Python para el manejo de estructuras de datos tabulares (DataFrames). 

A continuación, cargamos el dataset `housing` directamente desde un repositorio público.

In [9]:
# 1. Importamos la librería pandas. 
import pandas as pd

# 2. Le decimos dónde está el archivo. En este caso, el archivo está en internet, así que le damos la dirección URL.
url = 'https://raw.githubusercontent.com/ageron/handson-ml2/master/datasets/housing/housing.csv'

# 3. Leemos el archivo CSV y lo guardamos en una variable llamada 'df' (Data Frame).
df = pd.read_csv(url)

# 4. Imprimimos un mensaje para confirmar que los datos se han cargado correctamente.
print("Datos cargados")

Datos cargados


### 2. Visualización Inicial de los Datos
Para comprobar que el dataframe se ha construido correctamente, extraemos las primeras y las últimas. Esto nos permite familiarizarnos con la magnitud de los datos.

In [10]:
# Mostramos las 5 primeras filas para ver el encabezado
print("--- PRIMERAS 5 FILAS ---")
display(df.head())

# Mostramos las 5 últimas filas para comprobar el final del archivo
print("\n--- ÚLTIMAS 5 FILAS ---")
display(df.tail())

--- PRIMERAS 5 FILAS ---


,longitude,latitude,housing_median_age,total_rooms,total_bedrooms,population,households,median_income,median_house_value,ocean_proximity
0,-122.23,37.88,41.0,880.0,129.0,322.0,126.0,8.3252,452600.0,NEAR BAY
1,-122.22,37.86,21.0,7099.0,1106.0,2401.0,1138.0,8.3014,358500.0,NEAR BAY
2,-122.24,37.85,52.0,1467.0,190.0,496.0,177.0,7.2574,352100.0,NEAR BAY
3,-122.25,37.85,52.0,1274.0,235.0,558.0,219.0,5.6431,341300.0,NEAR BAY
4,-122.25,37.85,52.0,1627.0,280.0,565.0,259.0,3.8462,342200.0,NEAR BAY



--- ÚLTIMAS 5 FILAS ---


,longitude,latitude,housing_median_age,total_rooms,total_bedrooms,population,households,median_income,median_house_value,ocean_proximity
20635,-121.09,39.48,25.0,1665.0,374.0,845.0,330.0,1.5603,78100.0,INLAND
20636,-121.21,39.49,18.0,697.0,150.0,356.0,114.0,2.5568,77100.0,INLAND
20637,-121.22,39.43,17.0,2254.0,485.0,1007.0,433.0,1.7000,92300.0,INLAND
20638,-121.32,39.43,18.0,1860.0,409.0,741.0,349.0,1.8672,84700.0,INLAND
20639,-121.24,39.37,16.0,2785.0,616.0,1387.0,530.0,2.3886,89400.0,INLAND


 **Observaciones de la Visualización Inicial:**
Efectivamente, comprobamos que todas las columnas son numéricas (coordenadas, ingresos, edades) excepto la columna `ocean_proximity`, que contiene texto (ej. "NEAR BAY", "INLAND"). Esta columna requerirá codificación más adelante.

### 3. Tipos de Datos y Detección de Valores Nulos
Los algoritmos de Inteligencia Artificial requieren datos limpios y, por lo general, en formato numérico. Con el método `.info()` revisaremos qué tipo de dato contiene cada columna (números o texto). 

Además, es vital identificar si existen valores nulos (datos faltantes causados por errores de medición o de registro), ya que estos generarán errores en los algoritmos de Clustering que aplicaremos más adelante.

In [11]:
# Revisamos la estructura interna de la tabla y los tipos de variables
print("--- TIPOS DE DATOS POR COLUMNA ---")
df.info()

# Contamos cuántos huecos en blanco (nulos) hay en cada columna
print("\n--- RECUENTO DE VALORES NULOS ---")
display(df.isnull().sum())

--- TIPOS DE DATOS POR COLUMNA ---
<class 'pandas.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   longitude           20640 non-null  float64
 1   latitude            20640 non-null  float64
 2   housing_median_age  20640 non-null  float64
 3   total_rooms         20640 non-null  float64
 4   total_bedrooms      20433 non-null  float64
 5   population          20640 non-null  float64
 6   households          20640 non-null  float64
 7   median_income       20640 non-null  float64
 8   median_house_value  20640 non-null  float64
 9   ocean_proximity     20640 non-null  str    
dtypes: float64(9), str(1)
memory usage: 1.6 MB

--- RECUENTO DE VALORES NULOS ---


longitude               0
latitude                0
housing_median_age      0
total_rooms             0
total_bedrooms        207
population              0
households              0
median_income           0
median_house_value      0
ocean_proximity         0
dtype: int64

Se puede observar que el recuento de nulos revela que **faltan 207 datos en la columna `total_bedrooms`**.

### 4. Estadísticas Descriptivas
Finalmente, realizamos un resumen estadístico de las variables numéricas. Esto nos proporciona una visión rápida de los valores medios, las desviaciones estándar (dispersión de los datos) y los valores máximos y mínimos.

In [12]:
# Generamos el resumen estadístico
print("--- ESTADÍSTICA DESCRIPTIVA ---")
display(df.describe())

--- ESTADÍSTICA DESCRIPTIVA ---


,longitude,latitude,housing_median_age,total_rooms,total_bedrooms,population,households,median_income,median_house_value
count,20640.000000,20640.000000,20640.000000,20640.000000,20433.000000,20640.000000,20640.000000,20640.000000,20640.000000
mean,-119.569704,35.631861,28.639486,2635.763081,537.870553,1425.476744,499.539680,3.870671,206855.816909
std,2.003532,2.135952,12.585558,2181.615252,421.385070,1132.462122,382.329753,1.899822,115395.615874
min,-124.350000,32.540000,1.000000,2.000000,1.000000,3.000000,1.000000,0.499900,14999.000000
25%,-121.800000,33.930000,18.000000,1447.750000,296.000000,787.000000,280.000000,2.563400,119600.000000
50%,-118.490000,34.260000,29.000000,2127.000000,435.000000,1166.000000,409.000000,3.534800,179700.000000
75%,-118.010000,37.710000,37.000000,3148.000000,647.000000,1725.000000,605.000000,4.743250,264725.000000
max,-114.310000,41.950000,52.000000,39320.000000,6445.000000,35682.000000,6082.000000,15.000100,500001.000000


> **Conclusiones de la Estadística Descriptiva:**
Observamos diferencias de escala muy grandes entre las variables. Por ejemplo, `median_income` se mueve en valores de un solo dígito (media de 3.8), mientras que `total_rooms` llega a valores en los miles (media de 2635).

## 5. Limpieza y Preparación de Datos (Tratamiento de Nulos)
Para poder alimentar a los algoritmos de Clustering, vamos a realizar dos operaciones críticas de preprocesamiento:

1. **Tratamiento de Nulos:** Imputaremos los 207 valores faltantes de `total_bedrooms` utilizando la mediana de dicha columna, ya que la mediana es más robusta frente a valores atípicos (casas gigantes con muchas habitaciones frente a casas pequeñas con pocas) que la media aritmética.
2. **Codificación de Variables Categóricas:** La variable `ocean_proximity` contiene texto sin un orden jerárquico. Para evitar que la IA asuma que una ubicación "vale más" que otra, utilizaremos la técnica de *One-Hot Encoding* mediante la función `pd.get_dummies()` de Pandas, convirtiendo el texto en variables binarias (0 y 1).

In [13]:
# 1. TRATAMIENTO DE NULOS
mediana_habitaciones = df['total_bedrooms'].median()
df['total_bedrooms'] = df['total_bedrooms'].fillna(mediana_habitaciones)

# VISUALIZACIÓN ANTES DE CODIFICAR
print("--- 1. COLUMNA ORIGINAL (SIN SUSTITUIR) ---")
display(df[['ocean_proximity']].head())

print("\n--- 2. RECUENTO DE CASAS POR CLASE (TEXTO) ---")
# El método value_counts() nos dice cuántas casas pertenecen a cada categoría
display(df['ocean_proximity'].value_counts())


# 2. CODIFICACIÓN DE TEXTO A NÚMEROS (One-Hot Encoding con Pandas)
# Generamos las nuevas columnas de ceros y unos
df = pd.get_dummies(df, columns=['ocean_proximity'], dtype=int)


# --- VISUALIZACIÓN DESPUÉS DE CODIFICAR ---
print("\n--- 3. DATOS SUSTITUIDOS ---")
display(df.head())

print("\n--- 4. RECUENTO FINAL DE NULOS ---")
display(df.isnull().sum())

--- 1. COLUMNA ORIGINAL (SIN SUSTITUIR) ---


,ocean_proximity
0,NEAR BAY
1,NEAR BAY
2,NEAR BAY
3,NEAR BAY
4,NEAR BAY



--- 2. RECUENTO DE CASAS POR CLASE (TEXTO) ---


ocean_proximity
<1H OCEAN     9136
INLAND        6551
NEAR OCEAN    2658
NEAR BAY      2290
ISLAND           5
Name: count, dtype: int64


--- 3. DATOS SUSTITUIDOS ---


,longitude,latitude,housing_median_age,total_rooms,total_bedrooms,population,households,median_income,median_house_value,ocean_proximity_<1H OCEAN,ocean_proximity_INLAND,ocean_proximity_ISLAND,ocean_proximity_NEAR BAY,ocean_proximity_NEAR OCEAN
0,-122.23,37.88,41.0,880.0,129.0,322.0,126.0,8.3252,452600.0,0,0,0,1,0
1,-122.22,37.86,21.0,7099.0,1106.0,2401.0,1138.0,8.3014,358500.0,0,0,0,1,0
2,-122.24,37.85,52.0,1467.0,190.0,496.0,177.0,7.2574,352100.0,0,0,0,1,0
3,-122.25,37.85,52.0,1274.0,235.0,558.0,219.0,5.6431,341300.0,0,0,0,1,0
4,-122.25,37.85,52.0,1627.0,280.0,565.0,259.0,3.8462,342200.0,0,0,0,1,0



--- 4. RECUENTO FINAL DE NULOS ---


longitude                     0
latitude                      0
housing_median_age            0
total_rooms                   0
total_bedrooms                0
population                    0
households                    0
median_income                 0
median_house_value            0
ocean_proximity_<1H OCEAN     0
ocean_proximity_INLAND        0
ocean_proximity_ISLAND        0
ocean_proximity_NEAR BAY      0
ocean_proximity_NEAR OCEAN    0
dtype: int64

> **Conclusiones de la Limpieza y Codificación (Antes y Después):**
> * **Transformación:** La técnica de *One-Hot Encoding* (`pd.get_dummies`) ha funcionado. El texto ha desaparecido y, en su lugar, el dataset ha incorporado 5 nuevas columnas binarias al final. Ahora, si una casa está en una isla, tendrá un `1` en la columna `ocean_proximity_ISLAND` y un `0` en el resto.
> * **Dataset:** La comprobación final de nulos arroja un total de 0. Hemos convertido el dataset en una matriz 100% numérica, continua y sin huecos, cumpliendo el requisito estricto para poder aplicar la Inteligencia Artificial (K-Means y GMM) en el siguiente paso.